extmod/modlwip.c

```diff
@@ -1159,19 +1159,25 @@ STATIC mp_uint_t lwip_socket_ioctl(mp_obj_t self_in, mp_uint_t request, uintptr_
         uintptr_t flags = arg;
         ret = 0;
 
-        if (flags & MP_STREAM_POLL_RD && socket->incoming.pbuf != NULL) {
-            ret |= MP_STREAM_POLL_RD;
-        }
-
-        if (flags & MP_STREAM_POLL_WR && tcp_sndbuf(socket->pcb.tcp) > 0) {
-            ret |= MP_STREAM_POLL_WR;
-        }
+        switch (socket->state) {
+        case STATE_NEW:
+        case STATE_CONNECTING:
+            break;
```

Comment:
This will change existing behaviour: a new socket is writable and a connecting socket can also be polled.

Reply:
I've compiled the two branches:
Compiling master branch:
('flash    ', 35488)
('padding  ', 1376)
('irom0text', 542108)
('total    ', 578972)
('md5      ', 'b368e5b34b904155fb7bbaeb4286b89f')

Compiling with proposed fix:
('flash    ', 35488)
('padding  ', 1376)
('irom0text', 542124)
('total    ', 578988)
('md5      ', '2b8357a9fdc04b1ca118294922d8caba')

That adds 16 bytes, but the patch is clear, will return expected results for sockets in NEW, CONNECTING or CONNECTED states. Other states (which are some kind of error states) will return the desired poll flag. The comments in handling the STATE_PEER_CLOSED block suggests the same principle. Actually, it does break existing code where the socket was in an error state, now an application will not block on it indefinitely.

Would you prefer a change like this?
diff --git a/extmod/modlwip.c b/extmod/modlwip.c
index bbb01b5d..e1b0ad10 100644
--- a/extmod/modlwip.c
+++ b/extmod/modlwip.c
@@ -1163,7 +1163,7 @@ STATIC mp_uint_t lwip_socket_ioctl(mp_obj_t self_in, mp_uint_t request, uintptr_
             ret |= MP_STREAM_POLL_RD;
         }
 
-        if (flags & MP_STREAM_POLL_WR && tcp_sndbuf(socket->pcb.tcp) > 0) {
+        if (flags & MP_STREAM_POLL_WR && (socket->pcb.tcp == NULL || tcp_sndbuf(socket->pcb.tcp) > 0)) {
             ret |= MP_STREAM_POLL_WR;
         }

While this would definetily be a smaller change in code, I still prefer the other. As that handles connection states, it is more clear about what id does. Also I dont know what other cases might exist when pcb.tcp is not null but the socket is closed in any way.